core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1539,6 +1539,13 @@ abstract class RDD[T: ClassTag](
     // NOTE: we use a global lock here due to complexities downstream with ensuring
     // children RDD partitions point to the correct parent partitions. In the future
     // we should revisit this consideration.
+    if (doCheckpointCalled) {
+      throw new SparkException("Because job has been executed on this RDD, checkpoint won't work")
+    }
```

Comment:
I see what you mean, if I am not wrong, are you trying to deal with something like this ?

'''
val rdd1 = ...
val rdd2 = rdd1.filter()
val rdd3 = rdd2.map()

rdd1.checkpoint() 
rdd3.count()

rdd2.checkpoint()
rdd3.count()
'''

The latter will not cause rdd2 to get checkpointed since the earlier checkpoint already marked doCheckpointCalled = true ?


If yes, then it is an interesting case which might not have been considered - but I could be wrong.
+CC @mateiz which wrote the memorization of checkpoint ...